modules/gapi/src/backends/ocl/goclcore.cpp

```diff
@@ -523,6 +541,47 @@ GAPI_OCL_KERNEL(GOCLTranspose, cv::gapi::core::GTranspose)
     }
 };
 
+GAPI_OCL_KERNEL(GOCLBGR, cv::gapi::streaming::GBGR)
+{
+    static void run(const cv::MediaFrame& in, cv::UMat& out)
+    {
+        (void)in; (void)out;
+#ifdef HAVE_DIRECTX
+#ifdef HAVE_D3D11
+#ifdef HAVE_ONEVPL
+        auto d = in.desc();
+        GAPI_Assert(d.fmt == cv::MediaFormat::NV12);
+
+        // FIXME: consider a better solution.
+        // Current approach cannot be easily extended for other adapters (getHandle).
+        auto params = in.get<cv::gapi::wip::onevpl::VPLMediaFrameDX11Adapter>()->getHandle();
+        auto handle = cv::util::any_cast<mfxHDLPair>(params);
+        ID3D11Texture2D* texture = reinterpret_cast<ID3D11Texture2D*>(handle.first);
+        GAPI_Assert(texture != nullptr && "mfxHDLPair contains ID3D11Texture2D that is nullptr");
+
+        ID3D11Device* pD3D11Device = nullptr;
+        texture->GetDevice(&pD3D11Device);
+        GAPI_Assert(pD3D11Device != nullptr && "D3D11Texture2D::GetDevice returns pD3D11Device that is nullptr");
+
+        // FIXME: Initializing context in every run might degrade the performance
+        cv::ocl::Context& ctx = cv::directx::ocl::initializeContextFromD3D11Device(pD3D11Device);
+        GAPI_Assert(ctx.ptr() != nullptr && "Built without D3D11 support");
+
+        // FIXME: copying memory here!
+        // Not sure if it's possible without a copy here since
+        // there is no known way to convert ID3D11Texture2D to OCL memory (UMat),
+        // despite those 2 are both located (out is not allocated yet?) on the same GPU
+        cv::directx::convertFromD3D11Texture2D(texture, out);
+#else
+        GAPI_Assert(false && "Built without VPL support");
+#endif // HAVE_ONEVPL
+#else
+        GAPI_Assert(false && "Built without D3D11 support");
```

Comment:
I suggest to print human readable error in LOG

Reply:
Done